alpha/action/migrations/migrations.go

```diff
@@ -0,0 +1,121 @@
+package migrations
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+)
+
+type MigrationToken string
+
+const (
+	invalidMigration string = ""
+	NoMigrations     string = "none"
+	AllMigrations    string = "all"
+)
+
+type Migration interface {
+	Token() MigrationToken
+	Help() string
+	Migrate(*declcfg.DeclarativeConfig) error
+}
+
+func newMigration(token string, help string, fn func(config *declcfg.DeclarativeConfig) error) Migration {
+	return &simpleMigration{token: MigrationToken(token), help: help, fn: fn}
+}
+
+type simpleMigration struct {
+	token MigrationToken
+	help  string
+	fn    func(*declcfg.DeclarativeConfig) error
+}
+
+func (s simpleMigration) Token() MigrationToken {
+	return s.token
+}
+
+func (s simpleMigration) Migrate(config *declcfg.DeclarativeConfig) error {
+	return s.fn(config)
+}
+
+func (s simpleMigration) Help() string {
+	return s.help
+}
+
+type Migrations struct {
+	Migrations []Migration
+}
+
+func validateName(name string) (MigrationToken, error) {
+	if name == AllMigrations {
+		return MigrationToken(AllMigrations), nil
+	}
+	for _, migration := range allMigrations {
+		if migration.Token() == MigrationToken(name) {
+			return migration.Token(), nil
+		}
+	}
+	return MigrationToken(invalidMigration), fmt.Errorf("unknown migration level %q", name)
+}
+
+// allMigrations represents the migration catalog
+// the order of these migrations is important
+var allMigrations = []Migration{
+	newMigration(NoMigrations, "do nothing", func(_ *declcfg.DeclarativeConfig) error { return nil }),
+	newMigration("bundle-object-to-csv-metadata", `migrates bundles' "olm.bundle.object" to "olm.csv.metadata"`, bundleObjectToCSVMetadata),
+}
+
+func NewMigrations(name string) (*Migrations, error) {
+	if name == AllMigrations {
+		return &Migrations{Migrations: slices.Clone(allMigrations)}, nil
+	}
+
+	token, err := validateName(name)
+	if err != nil {
+		return nil, err
+	}
+
+	migrations := slices.Clone(allMigrations)
+
+	found := false
+	keep := migrations[:0]
+	for _, migration := range migrations {
+		keep = append(keep, migration)
+		if migration.Token() == token {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("unknown migration level %q", name)
+	}
```

Comment:
Now that we `validateName`, it seems like this is unreachable code? Should we drop this (or validateName)?

Reply:
Yeah, this is dead, but my intention was to throw in validateName temporarily to validate the tokenizing, and evidently forgot to replace.  I don't think we gain much efficiency from skipping one loop, but we'll have it now. 😉